relayer/chains/bitcoin/provider.go

```diff
@@ -0,0 +1,1437 @@
+			Height:   resultTx.Height,
+			Messages: []*relayTypes.Message{msg},
+		})
+	}
+	return messages, nil
+}
+
+func (p *Provider) getNumOfPipelines(diff int) int {
+	if diff <= runtime.NumCPU() {
+		return diff
+	}
+	return runtime.NumCPU() / 2
+}
+
+func (p *Provider) runBlockQuery(ctx context.Context, blockInfoChan chan *relayTypes.BlockInfo, fromHeight, toHeight uint64) uint64 {
+	done := make(chan bool)
+	defer close(done)
+
+	heightStream := p.getHeightStream(done, fromHeight, toHeight)
+
+	diff := int(toHeight-fromHeight) / 2
+	remain := (toHeight - fromHeight) % 2
+	if remain == 1 {
+		diff += 1
+	}
+
+	numOfPipelines := p.getNumOfPipelines(diff)
+	wg := &sync.WaitGroup{}
+	for i := 0; i < numOfPipelines; i++ {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, heightStream <-chan *HeightRange) {
+			defer wg.Done()
+			for heightRange := range heightStream {
+				blockInfo, err := p.fetchBlockMessages(ctx, heightRange)
+				if err != nil {
+					p.logger.Error("failed to fetch block messages", zap.Error(err))
+					continue
+				}
+				var messages []*relayTypes.Message
+				for _, block := range blockInfo {
+					messages = append(messages, block.Messages...)
+				}
+				blockInfoChan <- &relayTypes.BlockInfo{
+					Height:   heightRange.End,
+					Messages: messages,
+				}
+			}
+		}(wg, heightStream)
+	}
+	wg.Wait()
+	return toHeight
+}
+
+func (p *Provider) getAddressesFromTx(txOut *wire.TxOut, chainParams *chaincfg.Params) []string {
+	receiverAddresses := []string{}
+
+	scriptClass, addresses, _, err := txscript.ExtractPkScriptAddrs(txOut.PkScript, chainParams)
+	if err != nil {
+		p.logger.Error("Script: Unable to parse (possibly OP_RETURN)", zap.Error(err))
+	} else {
+		p.logger.Info("Script Class", zap.String("scriptClass", scriptClass.String()))
+		if len(addresses) > 0 {
+			p.logger.Info("Receiver Address", zap.String("address", addresses[0].String()))
+			receiverAddresses = append(receiverAddresses, addresses[0].String())
+		}
+	}
+
+	return receiverAddresses
+}
+
+// SubscribeMessageEvents subscribes to the message events
+// Expermental: Allows to subscribe to the message events realtime without fully syncing the chain
+func (p *Provider) SubscribeMessageEvents(ctx context.Context, blockInfoChan chan *relayTypes.BlockInfo, opts *types.SubscribeOpts, resetFunc func()) error {
+	return nil
+}
+
+// SetLastSavedHeightFunc sets the function to save the last saved height
+func (p *Provider) SetLastSavedHeightFunc(f func() uint64) {
+	p.LastSavedHeightFunc = f
+}
+
+// GetLastSavedHeight returns the last saved height
+func (p *Provider) GetLastSavedHeight() uint64 {
+	return p.LastSavedHeightFunc()
+}
+
+func (p *Provider) SetSerialNumberFunc(f func() *big.Int) {
+	p.LastSerialNumFunc = f
+}
+
+func (p *Provider) GetSerialNumber() *big.Int {
+	return p.LastSerialNumFunc()
+}
+
+func (p *Provider) FetchTxMessages(ctx context.Context, txHash string) ([]*relayTypes.Message, error) {
+	return nil, nil
```

Comment:
Add implementation here to fetch transaction based on any given txHash if possible as it would make manual replaying transaction much simpler and effective based on tx hash.